homeassistant/components/smartthings/__init__.py

```diff
@@ -0,0 +1,201 @@
+    from pysmartthings import SmartThings
+
+    if not hass.config.api.base_url.lower().startswith('https://'):
+        _LOGGER.warning("The 'base_url' of the 'http' component must be "
+                        "configured and start with 'https://'")
+        return False
+
+    api = SmartThings(async_get_clientsession(hass),
+                      entry.data[CONF_ACCESS_TOKEN])
+
+    remove_entry = False
+    try:
+        # See if the app is already setup. This occurs when there are
+        # installs in multiple SmartThings locations (valid use-case)
+        manager = hass.data[DOMAIN][DATA_MANAGER]
+        smart_app = manager.smartapps.get(entry.data[CONF_APP_ID])
+        if not smart_app:
+            # Validate and setup the app.
+            app = await api.app(entry.data[CONF_APP_ID])
+            smart_app = setup_smartapp(hass, app)
+
+        # Validate and retrieve the installed app.
+        installed_app = await validate_installed_app(
+            api, entry.data[CONF_INSTALLED_APP_ID])
+
+        # Get devices and their current status
+        devices = await api.devices(
+            location_ids=[installed_app.location_id])
+        await asyncio.gather(*[d.status.refresh() for d in devices])
+
+        # Setup device broker
+        broker = DeviceBroker(hass, devices,
+                              installed_app.installed_app_id)
+        broker.event_handler_disconnect = \
+            smart_app.connect_event(broker.event_handler)
+        hass.data[DOMAIN][DATA_BROKERS][entry.entry_id] = broker
+
+    except ClientResponseError as ex:
+        _LOGGER.exception(ex)
+        if ex.status in (400, 401, 403, 422):
+            remove_entry = True
+        else:
+            raise ConfigEntryNotReady from ex
+    except (ClientConnectionError, RuntimeWarning) as ex:
+        _LOGGER.exception(ex)
+        raise ConfigEntryNotReady from ex
+
+    if remove_entry:
+        hass.async_create_task(
+            hass.config_entries.async_remove(entry.entry_id))
+        # only create new flow if there isn't a pending one for SmartThings.
+        flows = hass.config_entries.flow.async_progress()
+        if not [flow for flow in flows if flow['handler'] == DOMAIN]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={'source': 'import'}))
+        return False
+
+    for component in SUPPORTED_PLATFORMS:
+        hass.async_create_task(hass.config_entries.async_forward_entry_setup(
+            entry, component))
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Unload a config entry."""
+    broker = hass.data[DOMAIN][DATA_BROKERS].pop(entry.entry_id, None)
+    if broker and broker.event_handler_disconnect:
+        broker.event_handler_disconnect()
+
+    for component in SUPPORTED_PLATFORMS:
+        hass.async_create_task(hass.config_entries.async_forward_entry_unload(
```

Comment:
No need to create task, instead you should await this.

```python
return all(await gather(forward_unload_tasks))
```

Reply:
Will update.